Huffman Chapter Completed.

jachiang opened the pull request:
Resolves: #54

jnewbery commented on 2.5-huffman.ipynb:
```diff
@@ -1,25 +1,207 @@
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    ">\\* If more than 2 leafs or nodes share the same weight during construction with the Huffman algorithm, the selection of the two nodes to generate the next parent becomes ambiguous. In our implementation, we first sort our node queue by `weight`, then by `tagged hash` value. This way, the sorting of the node queue is always unambiguous given a distinct set of leafs."
```
This is rendered as a quote. Is that what you intended?

jachiang replied:
Yes, but I doesn't look right, changed it to normal paragraph.

jnewbery commented on 2.5-huffman.ipynb:
```diff
@@ -1,25 +1,207 @@
+    "internal_pubkey.set(bytes.fromhex('03af455f4989d122e9185f8c351dbaecd13adca3eef8a9d38ef8ffed6867e342e3'))\n",
+    "\n",
+    "# Derive pay-to-pubkey tapleafs\n",
+    "privkeyA = ECKey().generate()\n",
```
replace all of these with `generate_key_pair()`

jnewbery commented on 2.5-huffman.ipynb:
```diff
@@ -1,25 +1,207 @@
+    "tapleafD = TapLeaf()\n",
+    "tapleafE = TapLeaf()\n",
+    "\n",
+    "tapleafA.construct_pk(pubkeyA)\n",
```
Can we make `construct_pk()` return `self` and make the construction look like:

```
tapleafA = Tapleaf().construct_pk(pubkeyA)
```

I think that's more compact and easier to understand.

jnewbery commented on 2.5-huffman.ipynb:
```diff
@@ -1,25 +1,207 @@
+    "print(\"Taptree descriptor: {}\\n\".format(taptree2.desc))\n",
+    "\n",
+    "segwit_v1_script2, tweak2, control_map2 = taptree2.construct()\n",
+    "print(\"TapTrees are identical: {}\".format(tweak == tweak2))\n"
```
nit: remove empty line

jnewbery commented on 2.5-huffman.ipynb:
```diff
@@ -1,25 +1,207 @@
+    "tapscripts = [(\"A\", tapleafA.script), (\"B\", tapleafB.script), (\"C\", tapleafC.script), (\"D\", tapleafD.script), (\"E\", tapleafE.script)]\n",
+    "\n",
+    "for n, script in tapscripts:\n",
+    "    # TODO: Print out heights of each tapscript.\n"
```
nit: remove empty line

jnewbery commented on 2.5-huffman.ipynb:
```diff
@@ -1,25 +1,207 @@
+    "\n",
+    "for n, script in tapscripts:\n",
+    "    # TODO: Print out heights of each tapscript.\n"
    ]
```
Suggestion: end chapter with a **congratulations** markdown cell, summarizing what has been learned.

jnewbery commented on solutions/2.5-huffman-solutions.ipynb:
```diff
@@ -0,0 +1,47 @@
+    "    controlblockA = control_map[script]\n",
+    "    height = len(controlblockA) - 33\n",
+    "    print(\"Tapscript {} is at height {}\".format(n, int(height/32)))\n",
+    "    "
```
nit: remove empty lines

jachiang commented:
Have incorporated @jnewbery feedback, thank you!

bitschmidty commented on 2.5-huffman.ipynb:
```diff
@@ -1,25 +1,208 @@
+    "tapleafD = TapLeaf().construct_pk(pubkeyD)\n",
+    "tapleafE = TapLeaf().construct_pk(pubkeyE)\n",
+    "\n",
+    "# Sorted queue: (5, A), (4, B), (3, C), (2, D), (1, E)\n",
```
Confused how the sorting and selecting works (see above), so I cannot comment further until I 'get it'

bitschmidty commented on solutions/2.5-huffman-solutions.ipynb:
```diff
@@ -0,0 +1,45 @@
+   "source": [
+    "tapscripts = [(\"A\", tapleafA.script), (\"B\", tapleafB.script), (\"C\", tapleafC.script), (\"D\", tapleafD.script), (\"E\", tapleafE.script)]\n",
+    "for n, script in tapscripts:\n",
+    "    controlblockA = control_map[script]\n",
```
I think just controlblock is ok since we are iterating and folks might get confused since we have an A TapLeaf

bitschmidty commented on 2.5-huffman.ipynb:
```diff
@@ -1,25 +1,208 @@
+    "Verify that the TapTree constructed in the examples above reflect the structure shown in the image at the beginning of the chapter.\n",
+    "\n",
+    "Reminder: The controlblock of a TapLeaf consists of: \n",
+    "* `[Tapscript Version + Internal Public Key] - 33B`\n",
```
I think these two items need more clarity. I am unclear on if you are telling me the sizes or doing some arithmetic.

jnewbery commented:
Thanks for addressing my feedback!

I think it'd be useful to explain more concretely what the weight function is. We're trying to minimize the expected size of the transaction, so I think it's something like:

**probability of script being used** x **total size of witness spending script**

Is that right?

jnewbery commented on 2.5-huffman.ipynb:
```diff
@@ -1,25 +1,208 @@
+   "source": [
+    "#### _Programming Exercise 2.5.3:_ Compute height of TapLeafs in TapTree\n",
+    "\n",
+    "Verify that the TapTree constructed in the examples above reflect the structure shown in the image at the beginning of the chapter.\n",
```
I don't think this exercise is testing/reinforcing the right thing. To answer it successfully, the student needs to know:

1. the structure of the `control_map` returned by `TapTree.construct()`
2. what a taproot control block is

We shouldn't be trying to teach (1) - the test framework should be something that we use to teach the concepts, not something we're trying to get the students to learn for its own sake. (2) isn't covered by this chapter, so shouldn't be tested here.

jachiang commented on 2.5-huffman.ipynb:
```diff
@@ -1,25 +1,208 @@
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "![test](images/huffman_intro0.jpg)"
```
The lowest weights are popped after each re-sort (I mistakenly wrote highest weight). So 546 is re-sorted to 456, 45 are popped and a parent node with weight 9 is created.

jachiang commented on 2.5-huffman.ipynb:
```diff
@@ -1,25 +1,208 @@
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "![test](images/huffman_intro0.jpg)"
```
Ok, will update.

jachiang commented on 2.5-huffman.ipynb:
```diff
@@ -1,25 +1,208 @@
    "metadata": {},
    "source": [
-    "![test](images/taptree3.jpg)"
+    "**Note:** If more than 2 leafs or nodes share the same weight during construction with the Huffman algorithm, the selection of the two nodes to generate the next parent becomes ambiguous. In our implementation, we first sort our node queue by `weight`, then by `tagged hash` value. This way, the sorting of the node queue is always unambiguous given a distinct set of leafs."
```
Hm, I suppose a standardized protocols will design output scripts with standard probability weights to optimize for tapscript spending cost (such as Lightning for example). I suspect there is a privacy trade-off we might want to mention, but would also like to get feedback from the BIP authors here: https://github.com/sipa/bips/issues/105#issue-509421166.

jachiang commented:
Thanks for the comments and improvements @bitschmidty @jnewbery. 

I have incorporated all your suggestions:
* Inline comments from bitschmidty.
* Better explanation of the goal: Minimize expected witness weight during script path spend.
* Replaced original exercise (height inferred from controlblock) with a probability weight assignment exercise which does not assume knowledge about controlblock.

jnewbery commented on test_framework/script.py:
```diff
@@ -1219,12 +1230,18 @@ def __init__(self, left=None, right=None):
         self.right = right
 
-    # TODO: Not nice, but seems to be necessary for sorting by priority queue.
+    def tagged_hash(self):
```
I think you can replace this whole function with something like:

```
def tagged_hash(self):
    # lexicographically sort the child hashes, concat them and hash with "TapBranch" tagged hash
    return TaggedHash("TapBranch", b''.join(sorted([h_l,h_r])))
```

jachiang replied:
Yes, this is much nicer. Thank you!

jnewbery commented on test_framework/script.py:
```diff
@@ -1096,6 +1106,7 @@ def from_desc(self, desc):
             self._decode_tree(desc[70:-1], parent=self.root)
         else:
             raise Exception
+        return self
```
What do you think about just allowing `__init__()` to take an optional descriptor?

jachiang replied:
I think its a good idea! Would it make sense to do this in another PR for all affected chapters? If you agree, I can track this in a dedicated issue.

jnewbery commented on 2.5-huffman.ipynb:
```diff
@@ -1,25 +1,248 @@
+    "\n",
+    "For example, for a TapTree with TapLeafs A, B, and C, the expected witness weight is determined as follows:\n",
+    "\n",
+    "* **`Expected-witness-weight =`**\n",
```
I think this looks better without the bullet points, for example:

```
Expected-witness-weight =
      Probability-of-A * Witness-weight-A
    + Probability-of-B * Witness-weight-B
    + Probability-of-C * Witness-weight-C
```

(ie use the backticks in your markdown to get fixed-width rendering)

jachiang replied:
Thanks for this, this is better. Fixed.

jnewbery commented on 2.5-huffman.ipynb:
```diff
@@ -1,25 +1,248 @@
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "**Note:** There may be commonly used Taproot output templates with known TapScripts and respective probabilities. Spending a TapScript will reveal both the leaf depth and TapScript, and may allow an observer to match this information against known standard output types. In this case, there exists a trade-off between spending cost and privacy."
```
I'd suggest dropping this. It's confusing and distracts from the main point of the chapter.

jachiang replied:
Ok. Removed.

jnewbery commented on 2.5-huffman.ipynb:
```diff
@@ -1,25 +1,248 @@
+   "source": [
+    "## Part 1: Huffman Algorithm\n",
+    "\n",
+    "We can use [Huffman's algorithm](https://en.wikipedia.org/wiki/Huffman_coding) to build tree structures from the leafs and their relative probabilities.\n",
```
s/leafs/leaves/

EDIT: I've just noticed that there are many instances of 'leafs' in this repo. We should fix those.

jachiang replied:
Thanks. Fixed.

jnewbery commented on 2.5-huffman.ipynb:
```diff
@@ -1,25 +1,248 @@
+    "    * **`+ Probability-of-B * Witness-weight-B`**\n",
+    "    * **`+ Probability-of-C * Witness-weight-C`**\n",
+    "\n",
+    "In order to minimize the expected witness weight for the script path, we can try to reduce the size of the required `inclusion proof` for those TapScripts which have a higher probability weighting by placing these closer to the root of the tree."
```
I think it's clearer if you replace 'probability weighting' with 'expected weight' here and throughout the chapter.

jachiang replied:
Thanks John, however, I am not sure I understand why it is clearer.

* The overall expected witness weight is the sum of witness weights * probability (based on probability distribution).
* The probability weighting captures the relative probability for a given leaf.

An "expected" value should be adjusted for probability-distribution, right? 
Shouldn't we reserve the term "expected" for this?

jnewbery replied:
Hmmm, yes perhaps you're right that _expected_ should only be used for the distribution. I think the most accurate term we're looking for is _probability-weighted weight_, but that sounds a bit clumsy.

I've just realised that you're assuming that the tapscript witness is the same weight for all tapscripts in this tree, and I think you're using the word 'weight' lower down in this notebook for 'probability'. Is that correct? If so, I think it's pretty confusing.

jnewbery commented on 2.5-huffman.ipynb:
```diff
@@ -1,25 +1,248 @@
+   ]
+  },
+  {
+   "cell_type": "code",
```
This exercise would be more effective if there was a clear pass/fail test at the bottom. I just entered random weights and there's no way for me to know whether I got the right answer or now.

jachiang replied:
Yes I agree. I have added a pass/fail test at the bottom which prints out heights of constructed tree. Cheers.

jnewbery commented:
I've merged the first two commits separately and rebased this on master.

jachiang commented on 2.5-huffman.ipynb:
```diff
@@ -1,25 +1,275 @@
+   "source": [
+    "# 2.5 Huffman TapTree Constructor\n",
+    "\n",
+    "When constructing a TapTree, there is the possibility of optimizing the tree structure so that the expected witness weight when spending along the script path is minimized. The `expected witness weight` is determined by summing up all probability-adjusted witness weights of the TapScripts in the TapTree.\n",
```
Thanks, updated with your suggestion.

jachiang commented on 2.5-huffman.ipynb:
```diff
@@ -1,25 +1,275 @@
+   "source": [
+    "# 2.5 Huffman TapTree Constructor\n",
+    "\n",
+    "When constructing a TapTree, there is the possibility of optimizing the tree structure so that the expected witness weight when spending along the script path is minimized. The `expected witness weight` is determined by summing up all probability-adjusted witness weights of the TapScripts in the TapTree.\n",
```
I Agree. Fixed.

jachiang commented on 2.5-huffman.ipynb:
````diff
@@ -1,25 +1,275 @@
+    "    + Probability-of-C * Witness-weight-C\n",
+    "```\n",
+    "\n",
+    "In order to minimize the expected witness weight for the script path, we can try to reduce the size of the required `inclusion proof` for those TapScripts which have a higher probability by placing these closer to the root of the tree."
````
Fixed.

jachiang commented:
> Hmmm, yes perhaps you're right that expected should only be used for the distribution. I think the most accurate term we're looking for is probability-weighted weight, but that sounds a bit clumsy.
> I've just realised that you're assuming that the tapscript witness is the same weight for all tapscripts in this tree, and I think you're using the word 'weight' lower down in this notebook for 'probability'. Is that correct? If so, I think it's pretty confusing.

Hm, I would suggest the following, let me know if you agree:
- "Assigned frequencies" (previously referred to as Huffman constructor _weights_).
- "Assigned" because they are arbitrary, it is their ratios which determine the tree structure.
- We can now reserve "weight" for witness/tx's.

Regarding tapscript witness weight for tapscripts:
- In the example all tapscripts have equal witness weights.
- Unless I missed something, they do not necessarily need to be equal, in order for Huffman to reduce the expected (probability-adjusted) witness weight.
- A byte saved in the highest probability tapscript spend has the highest savings impact on the overal expected witness weight, regardless of total witness size in bytes.
- Do you agree?

jachiang commented:
Updated based on @jnewbery's most recent review.

jnewbery commented:
Great stuff James. Thanks.

I've squashed in a couple of very small nits.
